pkg/ring/kv/memberlist/tcp_transport.go

```diff
@@ -396,9 +431,9 @@ func (t *TCPTransport) WriteTo(b []byte, addr string) (time.Time, error) {
 
 func (t *TCPTransport) writeTo(b []byte, addr string) error {
 	// Open connection, write packet header and data, data hash, close. Simple.
-	c, err := net.DialTimeout("tcp", addr, t.cfg.PacketDialTimeout)
+	c, err := t.getConnection(addr, t.cfg.PacketDialTimeout)
 	if err != nil {
-		return nil
```

Comment:
:grimacing: Any guesses how this would have manifested?

Reply:
It looks like any time there was a write operation that didn't succeed dialing, the client would assume it did. This can manifest itself in a lot of horrific ways but it's really hard to get a sense of how prevalent this issue is. It makes me question some of the behavior I've been observing w/ memberlist and some log lines that I hadn't dug into yet but assumed where inevitable network flake in a gossip system that could be sorted with tuning.

@pstibrany Just to verify, we didn't intend to return `nil` here for any reason? If not I'm also going to add a `BUGFIX` entry to the changelog.

Reply:
Oh, this is a good catch! If it was on purpose, I would expect some with explanation. Given it’s missing, I believe this is a bug.

Reply:
>  If not I'm also going to add a BUGFIX entry to the changelog.

Looking at the code I think the only impact would have been not counting the failure to `memberlist_tcp_transport_packets_sent_errors_total` metric, so I would mention it.

Also, while looking at this code, I've noticed the config option `ReportWriteToErrors` is always false, since it's not exposed through the config / CLI flag. What should we do with that?

Reply:
> Also, while looking at this code, I've noticed the config option ReportWriteToErrors is always false, since it's not exposed through the config / CLI flag. What should we do with that?

I created an issue to track this here: https://github.com/cortexproject/cortex/issues/4060